### PR TITLE
Feature/migrate eeprom to preferences

### PR DESCRIPTION
**Pull request by jMaddrell**
The current configuration approach makes it difficult to add new options, I have multiple KNOMI displays so currently they both try and use the "KNOMI" hostname.

This PR migrates configuration to the preferences library so that we can easily add new options, and adds a hostname option to the UI. 

After flashing "legacy configuration" is migrated by reading the EEPROM and then erasing and initialising the nvs partition.

During initial setup I have also disabled requests to the KlipperAPI server to clean up the logs.

The only issue I can identify is when connectToWiFi() is called when we exit ap mode dns.stop() can fail to free a buffer which causes the ESP to reset.

### CODE REVIEW

**Comment by jMaddrell** [comment]
In the future I'd like to add additional options such as the "base colour" that KNOMI uses for UI elements, and also the method of determining the print completion percentage (file-relative, file-absolute, slicer, filament usage).

File: Firmware/src/config/Config.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+void Config::save() {
+  preferences.begin(this->getConfigNamespace().c_str(), false);
+  preferences.putBool("initialised", this->isInitailised());
+  preferences.end();
+}
+
+void Config::reset() {
+  this->initialised = false;
```
Comment (jMaddrell) [inline Firmware/src/config/Config.cpp:28]: Is it worth making this do a full reset to defaults?

File: Firmware/src/network/WifiManager.h [hunk 2]
```diff
@@ -38,16 +38,21 @@ class WifiManager {
...
       ap = new WifiAccessPoint();
       return;
     }
     LV_LOG_INFO("We have config - let's try STA");
 
-    delete ap;
-    ap = nullptr;
-    sta = new WifiStation(config);
+    if (ap != nullptr) {
+      delete ap;
```
Comment (DiverOfDark) [inline Firmware/src/network/WifiManager.h:51]: delete on nullptr should do nothing - so check is an excessive AFAIK

Comment (jMaddrell) [reply Firmware/src/network/WifiManager.h]: Tested and push'd

File: Firmware/src/config/Config.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  void load();
+
+public:
+  Config() {
+    this->configNamespace = "knomi";
+    this->load();
+  }
+  NetworkConfig *getNetworkConfig() { return this->networkConfig; }
+  KlipperConfig *getKlipperConfig() { return this->klipperConfig; }
+  bool isInitailised() { return this->initialised; }
```
Comment (DiverOfDark) [inline Firmware/src/config/Config.h:27]: typo: isInit_AI_lised -> isInitialised

Comment (jMaddrell) [reply Firmware/src/config/Config.h]: Fixed

File: Firmware/src/config/Config.cpp (hunk 1 above)
Comment (jMaddrell) [reply Firmware/src/config/Config.cpp:28]: This now erases the nvs_partition to restore defaults and then restarts the esp

**Comment by DiverOfDark** [comment]
First of all - nice job!

I am not that sure of value of migration of legacy settings - it's not that much of them.  
At which point of time we will remove LegacySettings from codebase?

Also, I know that NVS is kinda standard settings storage - but have you considered saving settings into json file on littlefs?
Then it could be easily consumed or updated by web ui (and we wouldn't have limit of 20kb for all settings as we have currently) - we could have just common api for dl/upload of file. Downside of such approach could be lose of settings on FS flashing - so not an ideal stuff as well.

**Comment by jMaddrell** [comment]
In terms of migrating config I think it could be removed at any time, I just added it to make the migration smoother / pr getting approved easier.

I went with NVS to avoid configuration being cleared when the fs is updated, I don't see 20kb being much of a limitation for the configuration personally - for most of the settings I intend on adding I can always use a enum and just store the integer representation of it in NVS.

However if you'd prefer the config to be stored as json on the FS I'm happy to do that instead.

**Comment by DiverOfDark** [comment]
I don't have any strict preference on that part - be it nvs or file on fs - both have upsides and downsides - so let's keep it your current way, because it already works and there's no real requirement to change that.

Would be happy to merge PR once all outstanding comments are resolved.

**Comment by jMaddrell** [comment]
I’ve addressed all the comments and also made Config->reset() do a full reset of nvs which is only triggered by the button being held

**Pull request merged**